balance_strategy.go

```diff
@@ -127,3 +155,926 @@ func balanceStrategyHashValue(vv ...string) uint32 {
+			prevPartitions := make([]topicPartitionAssignment, 0)
+			for partition := range partitionsWithADifferentPreviousAssignment {
+				// from partitions that had a different consumer before, keep only those that are assigned to this consumer now
+				if memberAssignmentsIncludeTopicPartition(member.assignments, partition) {
+					prevPartitions = append(prevPartitions, partition)
+				}
+			}
+
+			if len(prevPartitions) > 0 {
+				// if there is a partition on this consumer that was assigned to another consumer before mark it as good options for reassignment
+				partition := prevPartitions[0]
+				prevPartitions = append(prevPartitions[:0], prevPartitions[1:]...)
+				member.assignments = removeTopicPartitionFromMemberAssignments(member.assignments, partition)
+				sortedPartitions = append(sortedPartitions, partition)
+				heap.Fix(&pq, member.index)
+			} else if len(member.assignments) > 0 {
+				// otherwise, mark any other one of the current partitions as a reassignment candidate
+				partition := member.assignments[0]
+				member.assignments = append(member.assignments[:0], member.assignments[1:]...)
+				sortedPartitions = append(sortedPartitions, partition)
+				heap.Fix(&pq, member.index)
+			} else {
+				heap.Remove(&pq, 0)
+			}
+		}
+
+		for partition := range partition2AllPotentialConsumers {
+			found := false
+			for _, p := range sortedPartitions {
+				if partition == p {
+					found = true
+					break
+				}
+			}
+			if !found {
+				sortedPartitions = append(sortedPartitions, partition)
+			}
+		}
+	} else {
+		// an ascending sorted set of topic partitions based on how many consumers can potentially use them
+		sortedPartitions = sortPartitionsByPotentialConsumerAssignments(partition2AllPotentialConsumers)
+	}
+	return sortedPartitions
+}
+
+func sortMemberIDsByPartitionAssignments(assignments map[string][]topicPartitionAssignment) []string {
+	// sort the members by the number of partition assignments in ascending order
+	sortedMemberIDs := make([]string, 0, len(assignments))
+	for memberID := range assignments {
+		sortedMemberIDs = append(sortedMemberIDs, memberID)
+	}
+	sort.SliceStable(sortedMemberIDs, func(i, j int) bool {
+		ret := len(assignments[sortedMemberIDs[i]]) - len(assignments[sortedMemberIDs[j]])
+		if ret == 0 {
+			return sortedMemberIDs[i] < sortedMemberIDs[j]
+		}
+		return len(assignments[sortedMemberIDs[i]]) < len(assignments[sortedMemberIDs[j]])
+	})
+	return sortedMemberIDs
+}
+
+func sortPartitionsByPotentialConsumerAssignments(partition2AllPotentialConsumers map[topicPartitionAssignment][]string) []topicPartitionAssignment {
+	// sort the members by the number of partition assignments in descending order
+	sortedPartionIDs := make([]topicPartitionAssignment, 0, len(partition2AllPotentialConsumers))
+	for partition := range partition2AllPotentialConsumers {
+		sortedPartionIDs = append(sortedPartionIDs, partition)
+	}
+	sort.SliceStable(sortedPartionIDs, func(i, j int) bool {
```

Comment:
The Java code here has three levels of comparison:
1) number of consumers, which is what is done here
2) topic name
3) partition
```
 872         public int compare(TopicPartition o1, TopicPartition o2) {
 873             int ret = map.get(o1).size() - map.get(o2).size();
 874             if (ret == 0) {
 875                 ret = o1.topic().compareTo(o2.topic());
 876                 if (ret == 0)
 877                     ret = o1.partition() - o2.partition();
 878             }
 879             return ret;
 880         }
```
This misses the latter two. `SliceStable` is unnecessary once they're added in.

I think that this also sorts from smallest to largest, rather than what you do here of largest to smallest:
```
 545             TreeSet<TopicPartition> sortedAllPartitions = new TreeSet<>(new PartitionComparator(partition2AllPotentialConsumers));
 546             sortedAllPartitions.addAll(partition2AllPotentialConsumers.keySet());
 547 
 548             while (!sortedAllPartitions.isEmpty())
 549                 sortedPartitions.add(sortedAllPartitions.pollFirst());
```
The tree is drained first to last.

FWIW, I'm writing this code up myself and trying to translate the Java code to something that actually makes sense. That's taking a long time. I'm pretty suspicious of some aspects of the Java code. I'm not 100% sure everything the Java code is doing is necessary (for example, I can't think of a way that one map in `areSubscriptionsIdentical` would return true while the other returns false).

However, since this PR is looking to be a direct transliteration, I'll try to vet the accuracy of that transliteration the best I can while I go through the Java and ... try to make sense of it.

Reply:
Note that the Java code is exceptionally confusing: if this is not a fresh assignment and all of the consumptions are identical, the sort order is by partitions with the most consumers to the partitions with the least consumers (and some prev consumers stuff).

In this else condition that does the sorting this comment chain is on, the Java code looks to sort by partitions with ascending number of consumers. Basically, the reverse of what it should be, solely because the Java code uses `pollFirst` instead of `pollLast`.

Am I reading that correctly?

Reply:
I am pretty confused about the sortPartitions function in general. It operates in two completely different ways: the normal way, which is ordering partitions by how many consumers can consume from them, and the has-a-lot-of-logic-way, which orders by partitions that were on the most loaded consumer.

The two completely different behaviors makes the future usage of the sorting hard to reason about :/